File: src/main/java/com/zepben/testutils/exception/ExpectException.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.testutils.exception;

@SuppressWarnings("WeakerAccess")
public class ExpectException {

    @FunctionalInterface
    public interface RunWithException {
        void run() throws Throwable;
    }

    private RunWithException func;

    public static ExpectException expect(RunWithException func) {
        return new ExpectException(func);
    }

    private ExpectException(RunWithException func) {
        this.func = func;
    }

    public ExceptionMatcher<Throwable> toThrow() throws ExpectExceptionError {
        try {
            func.run();
        } catch (Throwable e) {
            return new ExceptionMatcher<>(e);
        }
        throw new ExpectExceptionError("Throwable", "");
    }

    public <T extends Throwable> ExceptionMatcher<T> toThrow(Class<T> clazz) throws ExpectExceptionError {
        try {
            func.run();
        } catch (Throwable e) {
            if (clazz.isInstance(e))
                return new ExceptionMatcher<>(clazz.cast(e));

            throw new ExpectExceptionError(clazz.getSimpleName(), e.getClass().getSimpleName(), e);
        }
        throw new ExpectExceptionError(clazz.getSimpleName(), "");
    }

}

File: src/main/java/com/zepben/testutils/exception/ExpectException.kt
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.testutils.exception

class ExpectException private constructor(val func: RunWithException) {
    fun interface RunWithException {
        @Throws(Throwable::class)
        fun run()
    }

    @Throws(ExpectExceptionError::class)
    inline fun <reified T : Throwable> toThrow(): ExceptionMatcher<T> {
        try {
            func.run()
        } catch (e: Throwable) {
            if (e is T) return ExceptionMatcher(e)
            throw ExpectExceptionError(T::class.simpleName, e.javaClass.simpleName, e)
        }
        throw ExpectExceptionError(T::class.simpleName, "")
    }

    /**
     * Java-interoperable version of toThrow()
     */
    @Throws(ExpectExceptionError::class)
    fun <T : Throwable> toThrow(clazz: Class<T>): ExceptionMatcher<T> {
        try {
            func.run()
        } catch (e: Throwable) {
            if (clazz.isInstance(e)) return ExceptionMatcher(clazz.cast(e))
            throw ExpectExceptionError(clazz.simpleName, e.javaClass.simpleName, e)
        }
        throw ExpectExceptionError(clazz.simpleName, "")
    }

    @Throws(ExpectExceptionError::class)
    fun toThrowAny(): ExceptionMatcher<Throwable> = toThrow()

    companion object {
        @JvmStatic
        fun expect(func: RunWithException): ExpectException {
            return ExpectException(func)
        }
    }
}
